ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1892,12 +2069,332 @@ void shutdown() throws InterruptedException {
+        logger.trace("Thread name: {} stand by group {} is in iteration", threadName, standByGroupId);
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightGroups.get(standByGroupId).remoteReplicaInfos);
+      }
+
+      // filter active and stand by replicas
+      filterRemoteReplicasToReplicate(new ArrayList<>(nonIteratingNonPendingRemoteReplicas), activeReplicas,
+          standbyReplicasWithNoProgress);
+    }
+
+    /**
+     * Iterates over {@link #groupIdToRemoteReplicaMap} all potential groups,
+     * tries to create active groups and adds new groups to {@link #inflightGroups}
+     * also adds replicas which have timed out on stand by
+     * to {@link #standByReplicaQueue} for being picked by standby groups
+     */
+    private void createActiveGroups() {
+      logger.trace("Thread name: {} Trying to create new active groups", threadName);
+      for (Map.Entry<Integer, List<RemoteReplicaInfo>> entry : groupIdToRemoteReplicaMap.entrySet()) {
+        int groupId = entry.getKey();
+        logger.trace("Thread name: {} Processing for group id {}", threadName, groupId);
+        List<RemoteReplicaInfo> remoteReplicas = entry.getValue();
+        DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+        // get stand by group id for the data node
+        int standByGroupId = standByGroupIds.get(remoteNode);
+
+        List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+        // generate active and standby replicas with no progress
+        fillActiveAndStandByNoProgressReplicas(activeReplicas, standbyReplicasWithNoProgress, remoteReplicas,
+            standByGroupId, groupId);
+        logger.trace("Thread name: {} Active replicas are {}, standby Replicas with no progress are {}", threadName,
+            activeReplicas, standbyReplicasWithNoProgress);
+
+        // if the current group is not in iteration, and there are active replicas, we will start a new group
+        // and increase iteration count if we are able to create
+        if (!inflightGroups.containsKey(groupId)) {
+          if (!activeReplicas.isEmpty()) {
+            RemoteReplicaGroup groupForActiveReplicas =
+                new RemoteReplicaGroup(activeReplicas, remoteNode, false, groupId);
+            inflightGroups.put(groupId, groupForActiveReplicas);
+            groupIdIterationCountMap.put(groupId, groupIdIterationCountMap.getOrDefault(groupId, 0) + 1);
+            logger.trace("Thread name: {} Created new group for groupId {}, iteration count is {}", threadName, groupId,
+                groupIdIterationCountMap.getOrDefault(groupId, 0));
+          }
+        }
+
+        // get timed out stand by replicas with no progress
+        List<RemoteReplicaInfo> standByReplicasTimedOutOnNoProgress =
+            getRemoteStandbyReplicasTimedOutOnNoProgress(standbyReplicasWithNoProgress);
+
```

Comment:
There seems to be 3 types of replicas: active, standBy, standByTimedOut ? correct ?

First, can you please consistently get-rid of the term of NoProgress everywhere in your PR. I see its removed in some places where I pointed out but left in others. Please be consistent in the naming in all places.

Second, make a clear distinction between standBy and standByTimedOut replicas. 
For eg. this queue standByReplicaQueue is really a queue for timed-out replicas. correct ?